### PR TITLE
Add `Bitwarden.Server.Sdk`

### PR DESCRIPTION
**Pull request by justindbaur**
## 📔 Objective

This adds a `Bitwarden.Server.Sdk` project that is expected to be referenced via `<Sdk Name="Bitwarden.Server.Sdk" />` (the version will most likely be managed in `global.json`). Then you are able to dial in the exact feature set you want using properties. Say you don't want the logging this packages brings with it, you can say `<BitIncludeLogging>false</BitIncludeLogging>` and not only will it not register our logging code in startup it also won't include any of the dependencies such a feature requires. 

Right now, this is more of a proof of concept, I kept all the code in the `Bitwarden.Extensions.Hosting` project but we can start moving it into the `Bitwarden.Server.Sdk` project, when we move that over we need to add the `#if [FEATURE]` pragmas in various places to make all feature sets compilable. Right now, if you turn off any of the feature the compilation breaks. 

## ⏰ Reminders before review

- Contributor guidelines followed
- All formatters and local linters executed and passed
- Written new unit and / or integration tests where applicable
- Protected functional changes with optionality (feature flags)
- Used internationalization (i18n) for all UI strings
- CI builds passed
- Communicated to DevOps any deployment requirements
- Updated any necessary documentation (Confluence, contributing docs) or informed the documentation team

## 🦮 Reviewer guidelines

- 👍 (`:+1:`) or similar for great changes
- 📝 (`:memo:`) or ℹ️ (`:information_source:`) for notes or general info
- ❓ (`:question:`) for questions
- 🤔 (`:thinking:`) or 💭 (`:thought_balloon:`) for more open inquiry that's not quite a confirmed issue and could potentially benefit from discussion
- 🎨 (`:art:`) for suggestions / improvements
- ❌ (`:x:`) or ⚠️ (`:warning:`) for more significant problems or concerns needing attention
- 🌱 (`:seedling:`) or ♻️ (`:recycle:`) for future improvements or indications of technical debt
- ⛏ (`:pick:`) for minor or nitpick changes

### CODE REVIEW

File: extensions/Bitwarden.Server.Sdk/src/Bitwarden.Server.Sdk.csproj [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+
+    <PackageReadmeFile>./README.md</PackageReadmeFile>
+    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
+
+    <Version>0.1.10</Version>
+  </PropertyGroup>
+
+  <PropertyGroup>
+    <Authors>Bitwarden Inc.</Authors>
+    <Description>Bitwarden server sdk</Description>
```
Comment (withinfocus) [inline extensions/Bitwarden.Server.Sdk/src/Bitwarden.Server.Sdk.csproj:20]:
```suggestion
    <Description>Bitwarden Server SDK.</Description>
```

File: extensions/Bitwarden.Server.Sdk/src/README.md [hunk 2]
```diff
@@ -0,0 +1,8 @@
+# Bitwarden.Server.Sdk
+
+The Bitwarden server sdk is built for quickly getting started building
```
Comment (withinfocus) [inline extensions/Bitwarden.Server.Sdk/src/README.md:3]:
```suggestion
The Bitwarden Server SDK is built for quickly getting started building
```

File: extensions/Bitwarden.Server.Sdk/src/README.md [hunk 3]
```diff
@@ -0,0 +1,8 @@
...
+a Bitwarden flavored service. The entrypoint for using it is adding `UseBitwardenSdk()`
```
Comment (withinfocus) [inline extensions/Bitwarden.Server.Sdk/src/README.md:4]:
```suggestion
a Bitwarden-flavored service. The entrypoint for using it is adding `UseBitwardenSdk()`
```

File: extensions/Bitwarden.Server.Sdk/src/README.md [hunk 4]
```diff
@@ -0,0 +1,8 @@
...
+on your web application and configuring MSBuild properties to configure the features you
+want.
+
+## Feature Flags
```
Comment (withinfocus) [inline extensions/Bitwarden.Server.Sdk/src/README.md:8]: ❓ TODO?

File: extensions/Bitwarden.Server.Sdk/src/Bitwarden.Server.Sdk.csproj [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+    <!-- We are a meta package, we don't actually want people to have to use netstandard to use us. -->
+    <SuppressDependenciesWhenPacking>true</SuppressDependenciesWhenPacking>
+    <NoWarn>$(NoWarn);NU5128</NoWarn>
+
+    <IncludeBuildOutput>false</IncludeBuildOutput>
+
+    <PackageReadmeFile>./README.md</PackageReadmeFile>
+    <GeneratePackageOnBuild>true</GeneratePackageOnBuild>
+
+    <Version>0.1.10</Version>
```
Comment (withinfocus) [inline extensions/Bitwarden.Server.Sdk/src/Bitwarden.Server.Sdk.csproj:15]: 🎨 Need to get this working with CI so we can set it outside of code.

**Review by withinfocus — COMMENTED** [review]
What else needs to be done for CI / CD with this? Later PR?

File: extensions/Bitwarden.Server.Sdk/src/Bitwarden.Server.Sdk.csproj (hunk 5 above)
Comment (justindbaur) [reply extensions/Bitwarden.Server.Sdk/src/Bitwarden.Server.Sdk.csproj:15]: We will want it easily set in code. One of the downsides of SDKs is that they can only be consumed via nuget. So to test this I had to add a nuget source to the `bin/Release` directory of this project and had to make sure I incremented the version number in some way or I would have to clear my nuget cache everytime. 

I do personally like the version being part of the source control though, I'm not a big fan of setting it in CI.

File: extensions/Bitwarden.Server.Sdk/src/README.md (hunk 4 above)
Comment (justindbaur) [reply extensions/Bitwarden.Server.Sdk/src/README.md:8]: Yep, meant some brief documentation. Added it.

File: extensions/Bitwarden.Server.Sdk/src/README.md [hunk 6]
```diff
@@ -6,3 +6,13 @@ on your web application and configuring MSBuild properties to configure the feat
 want.
 
 ## Feature Flags
+
+Feature flag support can be added by adding `<BitIncludeFeatures>true</BitIncludeFeatures>` to
+your `csproj` file. The following API's become available:
```
Comment (withinfocus) [inline extensions/Bitwarden.Server.Sdk/src/README.md:11]:
```suggestion
your `csproj` file. The following APIs become available:
```

File: extensions/Bitwarden.Server.Sdk/src/README.md [hunk 7]
```diff
@@ -6,3 +6,13 @@ on your web application and configuring MSBuild properties to configure the feat
...
+
+- `IFeatureService` for getting values of features.
+- `RequireFeatureAttribute` for requiring a feature is enabled on controllers and controller actions.
+- `IEndpointConventionBuilder.RequireFeature()` for requiring a feature is enabled on minimal API's.
```
Comment (withinfocus) [inline extensions/Bitwarden.Server.Sdk/src/README.md:15]:
```suggestion
- `IEndpointConventionBuilder.RequireFeature()` for requiring a feature is enabled on minimal APIs.
```

File: extensions/Bitwarden.Server.Sdk/src/Bitwarden.Server.Sdk.csproj (hunk 5 above)
Comment (withinfocus) [reply extensions/Bitwarden.Server.Sdk/src/Bitwarden.Server.Sdk.csproj:15]: Alright, well I can get behind the version in code but do appreciate when we can control it via something that isn't an engineer pushing a commit. Maybe we copy what `server` does with a workflow? Have to think about how tags are made too, as that's what I am doing here now where an engineer pushes a tag and then we make a GitHub release by hand -- really should be all streamlined to me and I really want this repo to showcase best practices.

**Review by addisonbeck — APPROVED** [review]

**Pull request merged**